# Homework 1: Colin Recker

## Exercise 2.2: Altitude of a Satellite
A satellite is to be launched into a circular orbit around the Earth so that it orbits the planet
once every $T$ seconds.

**a)** Show that the altitude $h$ above the Earth's surface that the satellite must have is  
$$h=\left(\frac{GMT^2}{4\pi^2}\right)^{1/3}-R$$
where $G = 6.67 \times 10^{-11}$m$^3$kg$^{-1}$s$^{-2}$ is Newton's gravitational constant, $M=5.97 \times 10^{24}$ kg is the mass of the Earth, and $R=6371$ km is its radius.

**ANSWER:**
A satellite in circular orbit around Earth experiences a centripetal force described by the following:  
$$F_c = m\frac{v^2}{r}$$
This centripetal force is due to the gravitational force:  
$$F_g = \frac{GmM}{r^2}$$.  
Setting these equal to each other:  
$$
\begin{align}
m\frac{v^2}{r} &= \frac{GmM}{r^2} \\
    r &= \frac{GM}{v^2}
\end{align}
$$
The linear speed of the satellite is described in terms of the period and the circumference of its orbit $v=\frac{2\pi r}{T}$.  
Substituting this in for $v$:  
$$
\begin{align}
r &= \frac{GM}{\left(\frac{2\pi r}{T}\right )^2} \\
r^3 &= \frac{GMT^2}{4\pi^2} \\
r &= \left(\frac{GMT^2}{4\pi^2}\right)^{1/3}
\end{align}
$$  
Since r is the radial distance from the center of the Earth, $r$ is given by $r=h-R$
Thus, we find the result:  
$$h = \left(\frac{GMT^2}{4\pi^2}\right)^{1/3} - R $$

**b)**
Write a program that asks the user to enter the desired value of T and then calculates and
prints out the correct altitude in meters.

In [1]:
import numpy as np

In [2]:
#Defining constants
G = 6.67*(10**(-11)) #m^3 kg^-1 s^-2
M = 5.97*(10**(24)) #kg
R = 6371000 #in meters

In [3]:
#Altitude function
def h(T):
    return (G*M*(T**2)/(4*(np.pi**2)))**(1/3) - R

In [4]:
#Period input
T = float(input("Enter a value for a period T (seconds): "))
print(h(T), 'meters in altitude')

Enter a value for a period T (seconds): 100
-5905506.655312314 meters in altitude


**c)** 
Use your program to calculate the altitudes of satellites that orbit the Earth once a day
(so-called “geosynchronous” orbit), once every 90 minutes, and once every 45 minutes.
What do you conclude from the last of these calculations?

In [5]:
#90 Minutes
print(h(90*60), 'meters for 90 minute period')

#45 Minutes
print (h(45*60), 'meters for 45 minute period')

279321.62537285965 meters for 90 minute period
-2181559.8978108233 meters for 45 minute period


**ANSWER:** Shorter Periods imply smaller altitudes

**d)** Technically a geosynchronous satellite is one that orbits the Earth once per sidereal day,
which is 23.93 hours, not 24 hours. Why is this? And how much difference will it make
to the altitude of the satellite?  

**ANSWER:** The difference is due to the rotation of the Earth not matching it's orbit around the sun.

In [6]:
#Sidereal Day
print('For sidereal day (23.93 hours) the altitude is', h(23.93*3600), 'meters')
print('For 24-hr day:', h(24*3600), 'meters')
print('Difference in altitude:', h(24*3600)-h(23.93*3600), 'meters')

For sidereal day (23.93 hours) the altitude is 35773762.329895645 meters
For 24-hr day: 35855910.17617497 meters
Difference in altitude: 82147.8462793231 meters


## Exercise 2.3
Write a program to perform the inverse operation to that of Example 2.2. That
is, ask the user for the Cartesian coordinates x, y of a point in two-dimensional space, and
calculate and print the corresponding polar coordinates, with the angle $\theta$ given in degrees.

In [7]:
#Cartesian Inputs
x = float(input('x coordinate: '))
y = float(input('y coordinate: '))

#Conversion function
def cart_to_polar(x, y):
    #mapping x, y to r
    r = np.sqrt(x**2 + y**2)
    #mapping x, y to theta
    theta = np.arctan(y/x)
    #returning r and theta (in degrees)
    return r, np.degrees(theta)

r, theta = cart_to_polar(x, y)

print(f'The point ({x}, {y}) in polar:', f'({r}, {theta})')

x coordinate: 1
y coordinate: 1
The point (1.0, 1.0) in polar: (1.4142135623730951, 45.0)


## Exercise 2.5: Quantum Potential Step
A well-known quantum mechanics problem involves a particle of mass m that encounters a
one-dimensional potential step, like this:

![alt text](2.5.png)

The particle with initial kinetic energy $E$ and wavevector $k_1 = \frac{\sqrt{2mE}}{\hbar}$ enters from the left and encounters a sudden jump in potential energy of height $V$ at position $x = 0$. By solving the Schrödinger equation, one can show that when $E > V$ the particle may either (a) pass the step,
in which case it has a lower kinetic energy of $E − V$ on the other side and a correspondingly
smaller wavevector of $k_2 = \frac{\sqrt{2m(E − V)}}{\hbar}$, or (b) it may be reflected, keeping all of its kinetic energy and an unchanged wavevector but moving in the opposite direction. The probabilities
$T$ and $R$ for transmission and reflection are given by  
$T=\frac{4k_1k_2}{(k_1+k_2)^2}$, $R=\left(\frac{k_1-k_2}{k_1+k_2}\right)^2$  
Suppose we have a particle with mass equal to the electron mass $m = 9.11 \times 10^{−31}$ kg and
energy 10 eV encountering a potential step of height 9 eV. Write a Python program to compute
and print out the transmission and reflection probabilities using the formulas above.

In [8]:
#Constants
m = 9.11e-31 #kg
E = 10 #eV
V = 9 #eV
h_bar = 6.58e-16

In [9]:
#Calculating k_1 and k_2
k_1 = np.sqrt(2*m*E)/h_bar
k_2 = np.sqrt(2*m*(E-V))/h_bar

#Transmission/Reflection Probabilities
T = (4*k_1*k_2)/((k_1+k_2)**2)
R = ((k_1-k_2)/(k_1+k_2))**2

In [10]:
print(f'Transmission Probability: {T}')
print(f'Reflection Probability: {R}')

Transmission Probability: 0.7301261363877615
Reflection Probability: 0.26987386361223825


## Exercise 2.6: Planetary Orbits

The orbit in space of one body around another, such as a planet around the Sun, need not be circular. In general it takes the form of an ellipse, with the body sometimes closer in and sometimes further out. If you are given the distance $\ell_1$ of closest approach that a planet makes to the Sun, also called its *perihelion*, and its linear velocity $v_1$ at perihelion, then any other property of the orbit can be calculated from these two as follows.

**a)** Kepler’s second law tells us that the distance $\ell_2$ and velocity $v_2$ of the planet at its most
distant point, or aphelion, satisfy $\ell_2v_2 = \ell_1v_1$. At the same time the total energy, kinetic
plus gravitational, of a planet with velocity v and distance r from the Sun is given by  
$E=\frac{1}{2}mv^2-G\frac{mM}{r}$,  
where $m$ is the planet's mass, $M=1.98 \times 10^{30}$ kg is the mass of the Sun, and $G=6.6738 \times 10^{-11}$ m$^3$kg$^{-1}$s$^{-2}$ is Newton's gravitational constant. Given that energy must be conserved, show that $v_2$ is the smaller root of the quadratic equation  
$v_2^2-\frac{2GM}{v_1\ell_1}v_2-\left[v_1^2-\frac{2GM}{\ell_1}\right]=0$.  
Once we have $v_2$ we can calculate $\ell_2$ using the relation $\ell_2=\ell_1v_1/v_2$.

**ANSWER:**
First, we find the total energy at the aphelion ($\ell_1$) and perihelion($\ell_2$):
$E_1 = \frac{1}{2}mv_1^2-G\frac{mM}{\ell_1} \\
 E_2 = \frac{1}{2}mv_2^2-G\frac{mM}{\ell_2}$
By conservation of energy, we can set these equivalent to each other:  
$$
\begin{align}
    \frac{1}{2}mv_1^2-G\frac{mM}{\ell_1} &= \frac{1}{2}mv_2^2-G\frac{mM}{\ell_2} \\
    \frac{1}{2}v_1^2-\frac{1}{2}v_2^2 &= G\frac{M}{\ell_1}-G\frac{M}{\ell_2} \\
    v_1^2-v_2^2 &= \frac{2GM}{\ell_1}-\frac{2GM}{\ell_2} \\
    v_2^2 - v_1^2 + \frac{2GM}{\ell_1}-\frac{2GM}{\ell_2} &= 0
\end{align}
$$
By the relation $\ell_2=\ell_1v_1/v_2$,  
$$
\begin{align}
    v_2^2 - v_1^2 + \frac{2GM}{\ell_1}-\frac{2GM}{\frac{\ell_1v_1}{v_2}} &= 0 \\
    v_2^2 - \frac{2GM}{\ell_1v_1}v_2 - \left[v_1^2 - \frac{2GM}{\ell_1}\right] &= 0
\end{align}
$$


**b)** Given the values of $v_1$, $\ell_1$, and $\ell_2$, other parameters of the orbit are given by simple formulas can that be derived from Kepler's laws and the fact the orbit is an ellipse:  
Semi-major axis: $a=\frac{1}{2}(\ell_1+\ell_2)$  
Semi-minor axis: $b=\sqrt{\ell_1\ell_2}$  
Orbital period: $T=\frac{2\pi ab}{l_1v_1}$
Orbital eccentricity: $e=\frac{\ell_2-\ell_1}{\ell_2+\ell_1}$  
Test your program that asks the user to enter the distance to the Sun and velocity at perihelion, then calculates and prints the quantities $\ell_2$, $v_2$, $T$, and $e$.

In [11]:
#Constants/Inputs Definitions
M = 1.9891e30 #kg
G = 6.6738e-11 
v_1 = float(input("Enter v_1 (m/s): "))
l_1 = float(input("Enter l_1 (m): "))

Enter v_1 (m/s): 10
Enter l_1 (m): 10


In [12]:
#Quadratic formula function
def quadratic(a, b, c):
    positive = (-b+np.sqrt(b**2-4*a*c))/(2*a)
    negative = (-b-np.sqrt(b**2-4*a*c))/(2*a)
    return positive, negative

In [13]:
#Function that returns orbit properties
def orbit_props(v_1, l_1):
    v_2 = min(quadratic(1, (-2*G*M)/(v_1*l_1), -(v_1**2-(2*G*M/l_1))))
    l_2 = l_1*v_1/v_2
    a = 0.5*(l_1+l_2)
    b = np.sqrt(l_1*l_2)
    T = (2*np.pi*a*b)/(l_1*v_1)
    e = (l_2-l_1)/(l_2+l_1)
    print(f'l_2: {l_2} m\nv_2: {v_2} m/s\nT: {T} s\ne: {e}')
    return v_2, l_2, a, b, T, e

**c)** Test your program by having it calculate the properties of the orbits of the Earth (for
which $\ell_1 = 1.4710 × 10^{11}$ m and $v_1 = 3.0287 × 10^{4}$ ms$^{−1}$) and Halley’s comet ($\ell_1 =
8.7830 × 10^{10}$ m and $v_1 = 5.4529 × 10^{4}$ ms$^{−1}$). Among other things, you should find that
the orbital period of the Earth is one year and that of Halley’s comet is about 76 years.

In [14]:
#Earth's Orbital Properties:
print("Earth:") 
earth = orbit_props(3.0287e4, 1.471e11)
print(f"Earth Period (yr): {earth[4]/3600/24/365} \n")
#Halley's Comet
print("Halley's Comet:")
comet = orbit_props(5.4529e4, 8.783e10)
print(f"Halley's Comet Period (yr): {comet[4]/3600/24/365}")

Earth:
l_2: 152027197208.65994 m
v_2: 29305.39917726127 m/s
T: 31543060.207886923 s
e: 0.01647191313474219
Earth Period (yr): 1.0002238777234564 

Halley's Comet:
l_2: 5282214660876.441 m
v_2: 906.6806969191493 m/s
T: 2399312511.8451877 s
e: 0.9672889126454061
Halley's Comet Period (yr): 76.0817006546546


## Exercise 2.7: Catalan Numbers

The Catalan numbers Cn are a sequence of integers 1, 1, 2, 5, 14, 42, 132...that play an important
role in quantum mechanics and the theory of disordered systems. (They were central to Eugene
Wigner’s proof of the so-called semicircle law.) They are given by  
$C_0=1$, $C_{n+1}=\frac{4n+2}{n+2}C_n$  
Write a program that prints in increasing order all Catalan numbers less than or equal to one billion.

In [15]:
#Catalan Numbers <=1e9
n = 0
c = 1
while c < 1e9:
    c = int((4*n+2)/(n+2)*c)
    n += 1
    print(c)

1
2
5
14
42
132
429
1430
4862
16796
58786
208012
742900
2674440
9694845
35357670
129644790
477638700
1767263190


## A very useful trick!

Start with an empty list (r=\[\]). Create a while loop from MIN to MAX in steps of 1. Choose a function (sqrt, exponential, trig function, etc.). On each loop compute the value of that function (using the index; value = sqrt(MIN), sqrt(MIN+1), etc). Use the append function to add these values as elements of the list r (r.append(value)).

In [16]:
r = []
MIN = 0
MAX = 10
step = 1
while MIN!=MAX:
    r.append(round(np.sqrt(MIN), 4))
    MIN += step
print(r)

[0.0, 1.0, 1.4142, 1.7321, 2.0, 2.2361, 2.4495, 2.6458, 2.8284, 3.0]


## Exercise: Projectile motion
* Ask the user to input an initial velocity and launch angle, theta
*Use a while loop to compute the x and y coordinates of a projectile as a function of time.  
$x =x_0+v_{0x}t\text{       }$         $v_{0x}=v_{0x}\cos(\theta)$  
$y =y_0+v_{0x}t+0.5(9.8)t^2\text{       }$         $v_{0y}=v_{0y}\sin(\theta)$
*Try adjusting the time step for each iteration so it takes a few seconds to run the calculation. 
*Print data in three columns: t, x, y.

In [17]:
#Constants/Inputs
v_0 = float(input("Initial velocity: "))
theta = float(input("Launch angle: "))
x_0 = 0
y_0 = 0

def x(t):
    return x_0+v_0*np.cos(theta)*t
def y(t):
    return y_0+v_0*np.sin(theta)*t-0.5*(9.8)*(t**2)

#Time interval:
t = 0
t_max = 100
step = 0.001
print("t:\tx:\ty:")
while y(t) >= 0:
    print(f"{round(t, 3)}\t{round(x(t),2)}\t{round(y(t), 2)}")
    t+= step

Initial velocity: 10
Launch angle: 10
t:	x:	y:
0	0.0	0.0


## Exercise 2.8: Arrays and order of operations

Suppose arrays a and b are defined as follows:
```python
from numpy import array
a = array([1, 2, 3, 4], int)
b = array([1, 2, 3, 4], int)
```
What will the computer print upon executing the following lines? (Try to work out the answer before trying it on the computer.)

In [18]:
#Initializing arrays
from numpy import array
a = array([1, 2, 3, 4], int)
b = array([2, 4, 6, 8], int)

**a)**

In [19]:
print(b/a+1)

[3. 3. 3. 3.]


**b)**

In [20]:
print(b/(a+1))

[1.         1.33333333 1.5        1.6       ]


**c)**

In [21]:
print(1/a)

[1.         0.5        0.33333333 0.25      ]


## Exercise 2.9: The Madelung Constant
In condensed matter physics the Madelung constant gives the total electric potential felt by
an atom in a solid. It depends on the charges on the other atoms nearby and their locations.  
Consider for instance solid sodium chloride—table salt. The sodium chloride crystal has atoms
arranged on a cubic lattice, but with alternating sodium and chlorine atoms, the sodium ones
having a single positive charge +e and the chlorine ones a single negative charge −e, where e is
the charge on the electron. If we label each position on the lattice by three integer coordinates
(i, j, k), then the sodium atoms fall at positions where i + j + k is even, and the chlorine atoms
at positions where i + j + k is odd.
Consider a sodium atom at the origin, i = j = k = 0, and let us calculate the Madelung
constant. If the spacing of atoms on the lattice is a, then the distance from the origin to the atom
at position (i, j, k) is 
$\sqrt{(ia)^2+(jb)^2+(kc)^2}=a\sqrt{i^2+j^2+k^2}$  
and the potential at the origin created by such an atom is  
$V(i, j, k) = \pm\frac{e}{4\pi\epsilon_0a\sqrt{i^2+k^2+k^2}}$  
with $\epsilon_0$ being the permittivity of the vacuum and the sign of the expression depending on
whether $i + j + k$ is even or odd. The total potential felt by the sodium atom is then the sum of
this quantity over all other atoms. Let us assume a cubic box around the sodium at the origin,
with $L$ atoms in all directions. Then  
$V_{total} = \sum_{i,j,k=-L}^{L} V(i,j,k) = \frac{e}{4\pi\epsilon_0a}M$,  
where $M$ is the Madelung constant, at least approximately—technically the Madelung constant
is the value of $M$ when $L \rightarrow \infty$, but one can get a good approximation just by using a large
value of $L$.  
    Write a program to calculate and print the Madelung constant for sodium chloride. Use as
large a value of L as you can, while still having your program run in reasonable time—say in a
minute or less.

In [22]:
#Madelung Constant (Triple Sum)
L = 100
M = 0
for i in range(-L, L+1):
    for j in range(-L, L+1):
        for k in range(-L, L+1):
            if (i**2+j**2+k**2) == 0:
                pass
            elif (i+j+k)%2 == 0:
                M += (i**2+j**2+k**2)**(-1/2)
            else:
                M -= (i**2+j**2+k**2)**(-1/2)
print(M)

-1.7418198158396148


## Exercise 2.10 The Semi-Empirical Mass Formula
In nuclear physics, the semi-empirical mass formula is a formula for calculating the approx-
imate nuclear binding energy $B$ of an atomic nucleus with atomic number $Z$ and mass number $A$:  
$B=a_1A-a_2A^{2/3}-\frac{a_3Z^2}{A^{1/3}}-a_4\frac{(A-2Z)^2}{A}+\frac{a_5}{A^{1/2}}$, where, in units of millions of electron volts, the constants are $a_1 = 15.8$, $a_2 = 18.3$, $a_3 = 0.714$, $a_4 = 23.2$, and  
$$a_5 = \begin{cases}
    0 & \text{if $A$ is odd,} \\ 
    12.0  & \text{if $A$ and $Z$ are both even,} \\
    -12.0 & \text{if $A$ is even and $Z$ is odd.}
\end{cases}
$$

**a)** Write a program that takes as its input the values of $A$ and $Z$, and prints out the binding
energy for the corresponding atom. Use your program to find the binding energy of an
atom with $A = 58$ and $Z = 28$. (Hint: The correct answer is around 490 MeV.)

In [23]:
A = float(input("Enter A: "))
Z = int(input("Enter Z :"))
a_1 = 15.8 #MeV
a_2 = 18.3 #MeV
a_3 = 0.714 #MeV
a_4 = 23.2 #MeV

def BE(A, Z):
    if A%2 == 1:
        a_5 = 0
    elif Z%2 == 0:
        a_5 = 12.0
    else:
        a_5 = -12.0
    return (a_1*A)-a_2*(A**(2/3))-a_3*(Z**2)/(A**(1/3))-a_4*((A-(2*Z))**2)/(A)+a_5/(A**(1/2)) 
bind = BE(A, Z)
print(f"Binding energy: {bind} MeV")

Enter A: 58
Enter Z :28
Binding energy: 497.5620206224374 MeV


**b)**
Modify your program to print out not the total binding energy B, but the binding energy
per nucleon, which is B/A.

In [24]:
#Binding energy per nucleon
print(f"Binding energy per nucleon: {bind/A} MeV/nucleon")

Binding energy per nucleon: 8.578655527973059 MeV/nucleon


**c)**
Now modify your program so that it takes as input just a single value of the atomic
number $Z$ and then goes through all values of $A$ from $A = Z$ to $A = 3Z$, to find the one
that has the largest binding energy per nucleon. This is the most stable nucleus with the
given atomic number. Have your program print out the value of $A$ for this most stable
nucleus and the value of the binding energy per nucleon.

In [25]:
#Most stable nucleon
Z = int(input("Enter an atomic number Z: "))
def findBind(Z):
    energies_per = {}
    for A in range(Z, 3*Z+1):
        energies_per[A] = BE(A, Z)/A
    A = max(energies_per, key=energies_per.get)
    energy = max(energies_per.values())
    return A, energy
print(f"The highest energy is at A={findBind(Z)[0]} with energy {findBind(Z)[1]} MeV")

Enter an atomic number Z: 28
The highest energy is at A=62 with energy 8.70245768367189 MeV


**d)**
Modify your program again so that, instead of taking $Z$ as input, it runs through all
values of $Z$ from 1 to 100 and prints out the most stable value of $A$ for each one. At what
value of $Z$ does the maximum binding energy per nucleon occur? (The true answer, in
real life, is $Z = 28$, which is nickel.)

In [33]:
Zdict = {}
for Z in range(1, 101):
    Zdict[Z] = findBind(Z)
    A, energy_per = Zdict[Z]
    print(f"The highest energy/nucleon with Z={Z} is at A={A} with energy {energy_per} MeV/nucleon")
energyper = max([energy[1] for energy in list(Zdict.values())])
print(energyper)

The highest energy/nucleon with Z=1 is at A=3 with energy 0.36869091831015793 MeV/nucleon
The highest energy/nucleon with Z=2 is at A=4 with energy 5.321930578649441 MeV/nucleon
The highest energy/nucleon with Z=3 is at A=7 with energy 5.280168164356119 MeV/nucleon
The highest energy/nucleon with Z=4 is at A=8 with energy 6.466330085889912 MeV/nucleon
The highest energy/nucleon with Z=5 is at A=11 with energy 6.650123444727665 MeV/nucleon
The highest energy/nucleon with Z=6 is at A=14 with energy 7.200918138809924 MeV/nucleon
The highest energy/nucleon with Z=7 is at A=15 with energy 7.330860591990982 MeV/nucleon
The highest energy/nucleon with Z=8 is at A=18 with energy 7.719275577459026 MeV/nucleon
The highest energy/nucleon with Z=9 is at A=19 with energy 7.736977682756342 MeV/nucleon
The highest energy/nucleon with Z=10 is at A=22 with energy 8.035350864715019 MeV/nucleon
The highest energy/nucleon with Z=11 is at A=25 with energy 8.025554739665797 MeV/nucleon
The highest energy/nu

## Exercise 2.11: Binomial Coefficients
The binomial coefficient $\binom{n}{k}$ is an integer equal to  
$\binom{n}{k}=\frac{n!}{k!(n-k)!}=\frac{n\times(n-1)\times(n-2)\times...\times(n-k+1)}{1\times2\times...\times k}$  
when $k\ge1$, or $\binom{n}{0}=1$ when $k=0$.

**a)** 
Using this form for the binomial coefficient, write a user-defined function ```binomial(n,k)```
that calculates the binomial coefficient for given n and k. Make sure your function returns
the answer in the form of an integer (not a float) and gives the correct value of 1 for the
case where $k = 0$.

In [27]:
def binomial(n, k):
    numerator = np.math.factorial(n)
    denominator = np.math.factorial(k)*
    return int(np.math.factorial(n)/(np.math.factorial(k)*np.math.factorial(n-k)))
print(binomial(6, 2))

15


**b)**
Using your function write a program to print out the first 20 lines of “Pascal’s triangle.”
The nth line of Pascal’s triangle contains n + 1 numbers, which are the coefficients $\binom{n}{0}$,
$\binom{n}{1}$, and so on up to $\binom{n}{n}$. Thus the first few lines are  
1 1  
1 2 1  
1 3 3 1  
1 4 6 4 1

In [28]:
for n in range(1, 21):
    for k in range(n+1):
        print(binomial(n, k), end=' ')
    print()

1 1 
1 2 1 
1 3 3 1 
1 4 6 4 1 
1 5 10 10 5 1 
1 6 15 20 15 6 1 
1 7 21 35 35 21 7 1 
1 8 28 56 70 56 28 8 1 
1 9 36 84 126 126 84 36 9 1 
1 10 45 120 210 252 210 120 45 10 1 
1 11 55 165 330 462 462 330 165 55 11 1 
1 12 66 220 495 792 924 792 495 220 66 12 1 
1 13 78 286 715 1287 1716 1716 1287 715 286 78 13 1 
1 14 91 364 1001 2002 3003 3432 3003 2002 1001 364 91 14 1 
1 15 105 455 1365 3003 5005 6435 6435 5005 3003 1365 455 105 15 1 
1 16 120 560 1820 4368 8008 11440 12870 11440 8008 4368 1820 560 120 16 1 
1 17 136 680 2380 6188 12376 19448 24310 24310 19448 12376 6188 2380 680 136 17 1 
1 18 153 816 3060 8568 18564 31824 43758 48620 43758 31824 18564 8568 3060 816 153 18 1 
1 19 171 969 3876 11628 27132 50388 75582 92378 92378 75582 50388 27132 11628 3876 969 171 19 1 
1 20 190 1140 4845 15504 38760 77520 125970 167960 184756 167960 125970 77520 38760 15504 4845 1140 190 20 1 


**c)**
The probability that an unbiased coin, tossed n times, will come up heads k times is
$\binom{n}{k}$/$2n$. Write a program to calculate (a) the total probability that a coin tossed 100 times
comes up heads exactly 60 times, and (b) the probability that it comes up heads 60 or
more times.

In [35]:
#Probability a coin tossed 100 times comes up heads exactly 60 times
p_exact = binomial(100, 60)/(2**100)
print(f'Probability comes up heads exactly 60 times: {p_exact}')
p_more = 0
for k in range(60, 101):
    p_more += binomial(100, k)/(2**100)
print(f'Probability comes up head at least 60 times: {p_more}')

Probability comes up heads exactly 60 times: 0.010843866711637987
Probability comes up head at least 60 times: 0.028443966820490392


## Exercise 2.12: Prime Numbers
The program in Example 2.8 is not a very efficient way of calculating prime numbers: it checks
each number to see if it is divisible by any number less than it. We can develop a much faster
program for prime numbers by making use of the following observations:  
  
a) A number n is prime if it has no prime factors less than n. Hence we only need to check
if it is divisible by other primes.  
  
b) If a number n is non-prime, having a factor r, then n = rs, where s is also a factor. If
r ≥ √n then n = rs ≥ √ns, which implies that s ≤ √n. In other words, any non-
prime must have factors, and hence also prime factors, less than or equal to √n. Thus
to determine if a number is prime we have to check its prime factors only up to and
including √n—if there are none then the number is prime.  
  
c) If we find even a single prime factor less than √n then we know that the number is non-
prime, and hence there is no need to check any further—we can abandon this number
and move on to something else.  
  
Write a Python program that finds all the primes up to ten thousand. Create a list to store
the primes, which starts out with just the one prime number 2 in it. Then for each number n
from 3 to 10 000 check whether the number is divisible by any of the primes in the list up to
and including $\sqrt{n}$. As soon as you find a single prime factor you can stop checking the rest of
them—you know $n$ is not a prime. If you find no prime factors $\sqrt{n}$ or less then $n$ is prime and
you should add it to the list. You can print out the list all in one go at the end of the program,
or you can print out the individual numbers as you find them.

In [30]:
#Initial prime list
primes = [2]
for n in range(3, 10001):
    comp = False
    while not comp:
        for prime in primes:
            if prime > np.sqrt(n):
                break
            elif n%prime == 0:
                comp = True
                break
        break
    if not comp: 
        primes.append(n)
print(primes)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997, 1009, 1013, 1019, 1021, 1031, 1033, 1039, 1049, 1051, 1061, 1063, 1069, 1087, 1091, 1093, 1097, 1103, 1109, 1117, 1123, 1129, 1151, 1153, 1163, 1171, 1181, 1187, 1193, 1201, 1213, 1217, 12

## Exercise 2.13 Recursion
A useful feature of user-defined functions is *recursion*, the ability of a function to call itself. For
example, consider the following definition of the factorial $n!$ of a positive integer $n$:
$$
n! = \begin{cases}
    1 & \text{if }n=1, \\ 
    n \times (n-1)!  & \text{if }n>1
\end{cases}
$$  
This constitutes a complete definition of the factorial which allows us to calculate the value of
$n!$ for any positive integer. We can employ this definition directly to create a Python function
for factorials, like this:  
```python
def factorial(n):
    if n==1:
        return 1
    else:
        return n*factorial(n-1)
```
Note how, if $n$ is not equal to 1, the function calls itself to calculate the factorial of $n-1$. This is recursion. If we now say "```print(factorial(5))```" the computer will correctly print the answer 120.

**a)**
We encountered the Catalan numbers Cn previously in Exercise 2.7 on page 46. With just
a little rearrangement, the definition given there can be rewritten in the form

$$
C_n = \begin{cases}
    1 & \text{if }n=0, \\ 
    \frac{4n-2}{n+1}C_{n-1} & \text{if }n>0.
\end{cases}
$$  
Write a Python function, using recursion, that calculates $C_n$. Use your function to calcu-
late and print $C_{100}$.

In [31]:
#Catalan numbers: recursive function
def catalan(n):
    if n == 0:
        return 1
    return int((4*n-2)/(n+1)*catalan(n-1))
print(catalan(100))

896519947090131678185430741499021921911623568450622849024


**b)**
Euclid showed that the greatest common divisor $g(m, n)$ of two nonnegative integers $m$
and $n$ satisfies
$$
g(m,n) = \begin{cases}
    m & \text{if }n=0, \\ 
    g(n, m\text{ mod }n) & \text{if }n>0.
\end{cases}
$$  
Write a Python function $g(m,n)$ that employs recursion to calculate the greatest common
divisor of $m$ and $n$ using this formula. Use your function to calculate and print the greatest common divisor of 108 and 192.

In [32]:
#Greatest common divisor: recursive function
def g(m, n):
    if n == 0:
        return m
    return g(n, m%n)
print(g(108, 192))

12


Comparing the calculation of the Catalan numbers in part (a) above with that of Exercise 2.7,
we see that it’s possible to do the calculation two ways, either directly or using recursion. In
most cases, if a quantity can be calculated *without* recursion, then it will be faster to do so, and
we normally recommend taking this route if possible. There are some calculations, however,
that are essentially impossible (or at least much more difficult) without recursion. We will see
some examples later in this book.